src/proj.rs

```diff
@@ -0,0 +1,791 @@
+    )
+where
+    P: Pixel + 'static,
+    <P as Pixel>::Subpixel: Send + Sync,
+    <P as Pixel>::Subpixel: ValueInto<f32> + Clamp<f32>,
+    Fc: Fn(f32, f32) -> (f32, f32) + Send + Sync,
+    Fi: Fn(f32, f32) -> P + Send + Sync,
+{
+    let width = out.width();
+    let raw_out = out.as_mut();
+    let pitch = P::CHANNEL_COUNT as usize * width as usize;
+
+    raw_out.par_chunks_mut(pitch).enumerate().for_each(|(y, row)| { 
+        for (x, slice) in row.chunks_mut(P::CHANNEL_COUNT as usize).enumerate() {
+            let (px, py) = coord_tf(x as f32, y as f32);
+            *P::from_slice_mut(slice) = get_pixel(px, py);
+        }
+    });
+        
+}
+
+fn class_from_matrix(mx: [f32; 9]) -> TransformationClass {
+
+    if (mx[6]-0.0).abs() < 1e-10 &&
+       (mx[7]-0.0).abs() < 1e-10 &&
+       (mx[8]-1.0).abs() < 1e-10
+    {  
+        if (mx[0]-1.0).abs() < 1e-10 &&
+           (mx[1]-0.0).abs() < 1e-10 &&
+           (mx[3]-0.0).abs() < 1e-10 &&
+           (mx[4]-1.0).abs() < 1e-10 
+        {
+            TransformationClass::Translation 
+        } else { 
+            TransformationClass::Affine
+        }
+    } else {
+        TransformationClass::Projection
+    }
+}
+
+fn normalize(mx: [f32; 9]) -> [f32; 9] {
+    [mx[0]/mx[8], mx[1]/mx[8], mx[2]/mx[8],
+     mx[3]/mx[8], mx[4]/mx[8], mx[5]/mx[8],
+     mx[6]/mx[8], mx[7]/mx[8], mx[8]/mx[8]]
+}
+
+fn try_inverse(t: &[f32; 9]) -> Option<[f32; 9]> {
+    let (
+        t00, t01, t02,
+        t10, t11, t12,
+        t20, t21, t22
+    ) = (
+        t[0], t[1], t[2],
+        t[3], t[4], t[5],
+        t[6], t[7], t[8]
+    );
+
+    let m00 = t11 * t22 - t12 * t21;
+    let m01 = t10 * t22 - t12 * t20;
+    let m02 = t10 * t21 - t11 * t20;
+
+    let det = t00 * m00 - t01 * m01 + t02 * m02;
+
+    if (det).abs() < 1e-10 {
+        return None;
+    }
+
+    let m10 = t01 * t22 - t02 * t21;
+    let m11 = t00 * t22 - t02 * t20;
+    let m12 = t00 * t21 - t01 * t20;
+    let m20 = t01 * t12 - t02 * t11;
+    let m21 = t00 * t12 - t02 * t10;
+    let m22 = t00 * t11 - t01 * t10;
+
+    let inv = [
+        m00 / det, -m10 / det,  m20 / det,
+       -m01 / det,  m11 / det, -m21 / det,
+        m02 / det, -m12 / det,  m22 / det
+    ];
+
+    Some(normalize(inv))
+}
+
+fn mul3x3(a: [f32; 9], b: [f32; 9]) -> [f32; 9] {
+    let (
+        a11, a12, a13,
+        a21, a22, a23,
+        a31, a32, a33
+        ) = (
+        a[0], a[1], a[2],
+        a[3], a[4], a[5], 
+        a[6], a[7], a[8]
+        );
+    let (
+        b11, b12, b13,
+        b21, b22, b23,
+        b31, b32, b33
+        ) = (
+        b[0], b[1], b[2],
+        b[3], b[4], b[5], 
+        b[6], b[7], b[8]
+        );
+
+    [a11*b11 + a12*b21 + a13*b31, a11*b12 + a12*b22 + a13*b32, a11*b13 + a12*b23 +a13*b33,
+     a21*b11 + a22*b21 + a23*b31, a21*b12 + a22*b22 + a23*b32, a21*b13 + a22*b23 + a23*b33,
+     a31*b11 + a32*b21 + a33*b31, a31*b12 + a32*b22 + a33*b32, a31*b13 + a32*b23 + a33*b33]
+}
+
+fn blend<P>(
```

Comment:
This is copy-pasted without modification from `affine`, right? Reuse that function here.

Reply:
Yup. Didn't want to make it public in the affine module, and I was hoping to replace it with something faster, but couldn't achieve better perf with a couple of simd crates I have tried.